batcher/sweep_batch.go

```diff
@@ -0,0 +1,1145 @@
+		chainNotifier:    batchKit.chainNotifier,
+		signerClient:     batchKit.signerClient,
+		muSig2SignSweep:  batchKit.musig2SignSweep,
+		store:            batchKit.store,
+		Cfg:              &cfg,
+	}, nil
+}
+
+// NewBatch creates a new batch that already existed in storage.
+func NewBatchFromDB(cfg BatchConfig, batchKit BatchKit) (*Batch, error) {
+
+	newID := lntypes.Hash{}
+	_, err := crand.Read(newID[:])
+	if err != nil {
+		return nil, err
+	}
+
+	return &Batch{
+		ID:               batchKit.ID,
+		State:            batchKit.state,
+		PrimarySweepID:   batchKit.primaryID,
+		Sweeps:           batchKit.sweeps,
+		RequestChan:      make(chan *Sweep),
+		returnChan:       batchKit.returnChan,
+		blockEpochChan:   make(chan int32),
+		spendChan:        make(chan *chainntnfs.SpendDetail),
+		spendRefreshChan: make(chan int32, 1),
+		confChan:         make(chan *chainntnfs.TxConfirmation),
+		confRefreshChan:  make(chan int32, 1),
+		errChan:          make(chan error),
+		completeChan:     batchKit.completeChan,
+		BatchTxid:        batchKit.batchTxid,
+		BatchPkScript:    batchKit.batchPkScript,
+		RbfCache:         batchKit.rbfCache,
+		wallet:           batchKit.wallet,
+		chainNotifier:    batchKit.chainNotifier,
+		signerClient:     batchKit.signerClient,
+		muSig2SignSweep:  batchKit.musig2SignSweep,
+		store:            batchKit.store,
+		Cfg:              &cfg,
+	}, nil
+}
+
+// addSweep adds a sweep to the batch. If this is the first sweep being added
+// to the batch then it also sets the primary sweep ID.
+func (b *Batch) addSweep(ctx context.Context, sweep Sweep) error {
+
+	if b.PrimarySweepID == lntypes.ZeroHash {
+		b.PrimarySweepID = sweep.SwapHash
+		b.Cfg.BatchConfTarget = sweep.ConfTarget
+	}
+
+	if b.PrimarySweepID == sweep.SwapHash {
+		b.Cfg.BatchConfTarget = sweep.ConfTarget
+	}
+
+	if !b.SweepExists(sweep.SwapHash) {
+		log.Infof("Batch %x: adding sweep %x", b.ID[:6],
+			sweep.SwapHash[:6])
+	}
+
+	b.Sweeps[sweep.SwapHash] = sweep
+
+	err := b.persistSweep(ctx, sweep)
+	if err != nil {
+		log.Errorf("Batch %x error while persisting: %v", b.ID[:6],
+			err)
+	}
+
+	return err
+}
+
+// SweepExists returns true if the batch contains the sweep with the given hash.
+func (b *Batch) SweepExists(hash lntypes.Hash) bool {
+	_, ok := b.Sweeps[hash]
+	return ok
+}
+
+// AcceptsSweep returns true if the batch is able to accept the given sweep. To
+// accept a sweep a batch needs to be in an Open state, and the incoming sweep's
+// timeout must not be too far away from the other timeouts of existing sweeps.
+func (b *Batch) AcceptsSweep(sweep *Sweep) bool {
+	tooFar := false
+	for _, s := range b.Sweeps {
+		if int32(math.Abs(float64(sweep.Timeout-s.Timeout))) >
+			b.Cfg.MaxTimeoutDistance {
+
+			tooFar = true
+		}
+	}
+
+	return (b.State == Open) && !tooFar
+}
+
+// Run is the batch's main event loop.
+func (b *Batch) Run(ctx context.Context) error {
+	// Dispatch the blockheight monitor.
+	go b.monitorBlockHeight(ctx)
```

Comment:
I made it into a goroutine only to keep the `Run` body simpler / easier to read

Reply:
I think it is generally better to keep execution linearized when possible. Block notification happens every 10 minutes on average which is a rare event so no need for another goroutine.

Reply:
Agree, but this would also further extend the main `Run` method's select cases, as now I would also have to add the blockNtfnErr chan to it

Reply:
> Maybe having it in one place is still more readable.

Could you please clarify whether you mean you prefer hiding it behing `go b.monitorHeight(ctx)` or not?

My last argument for the goroutine would be that the resource cost and overhead, especially for low-frequency stuff (like getting a new height once every 10m), is so low that it's unreasonable to argue on the performance impact.

I personally find it simpler when I don't blow-up too many codelines in the main function of a struct as it can be visual noise. Although that's subjective, so np getting rid of the routine